### PR TITLE
Upgrade CodeMirror to version 6

### PR DESCRIPTION
**Pull request by mjfwebb**
# Description

This PR introduces CodeMirror version 6. The major change is how we interact with CodeMirror state, which requires us to use a bit more complex change handling when we consider our need to have multiple views, multiple tabs, and a query explorer.

## Complexity

Medium

### CODE REVIEW

File: packages/graphql-toolbox/package.json [hunk 1]
```diff
@@ -42,14 +42,18 @@
...
         "@graphiql/react": "0.17.6",
         "@neo4j-ndl/base": "1.5.1",
         "@neo4j-ndl/react": "1.5.1",
         "@neo4j/graphql": "3.20.1",
         "@neo4j/introspector": "1.0.3",
         "classnames": "2.3.2",
-        "codemirror": "5.65.13",
-        "codemirror-graphql": "2.0.8",
+        "cm6-graphql": "0.0.8",
+        "codemirror": "^6.0.1",
```
Comment (tbwiss) [inline packages/graphql-toolbox/package.json:56]:
```suggestion
        "codemirror": "6.0.1",
```

File: packages/graphql-toolbox/package.json [hunk 2]
```diff
@@ -42,14 +42,18 @@
     },
     "author": "Neo4j",
     "dependencies": {
+        "@codemirror/autocomplete": "6.7.1",
+        "@codemirror/commands": "6.2.4",
+        "@codemirror/lang-javascript": "^6.1.8",
```
Comment (tbwiss) [inline packages/graphql-toolbox/package.json:47]:
```suggestion
        "@codemirror/lang-javascript": "6.1.8",
```

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-3468.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

File: packages/graphql-toolbox/src/modules/EditorView/Editor.tsx [hunk 3]
```diff
@@ -165,14 +296,13 @@ export const Editor = ({ schema }: Props) => {
...
-                                        schema={schema}
                                         loading={loading}
-                                        mirrorRef={refForQueryEditorMirror}
-                                        executeQuery={onSubmit}
-                                        query={useStore.getState().getActiveTab().query}
-                                        onChangeQuery={(query) => {
-                                            store.updateQuery(query, useStore.getState().activeTabIndex);
-                                        }}
+                                        elementRef={elementRef}
+                                        // onChangeQuery={(query) => {
```
Comment (tbwiss) [inline packages/graphql-toolbox/src/modules/EditorView/Editor.tsx:301]: Shall we remove these comments?

**Comment by neo4j-team-graphql** [comment]
# Performance Report

| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
| 🟥 query.QueryWithNestedIn | 14744 | 13190 | 68 | 82 | 1664 |


🟥 - Performance worsened (dbHits)
🟩 - Performance improved (dbHits)
🟦 - New test
        

<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 65 | 54 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 134 | 139 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 100 | 65 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 3939 | 4144 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 3468 | 3522 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 47 | 57 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 51 | 62 | 2174 |
|  aggregations.DeeplyNestedCount | 13070331 | 13070331 | 5412 | 5839 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 188 | 176 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 75 | 77 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 324 | 341 | 3003 |
|  connections.Connection | 14082 | 14082 | 68 | 86 | 2174 |
|  connections.NestedConnection | 45472 | 45472 | 144 | 157 | 4516 |
|  connections.ConnectionWithSort | 3284 | 3284 | 81 | 120 | 1040 |
|  connections.ConnectionWithSortAndCypher | 3284 | 3284 | 257 | 421 | 1040 |
|  create.SimpleMutation | 7 | 7 | 71 | 80 | 1 |
|  cypher-directive.TopLevelCypherDirective | 122100 | 122100 | 706 | 827 | 12241 |
|  cypher-directive.TopLevelCypherDirectiveWithColumnName | 153617 | 153617 | 208 | 262 | 12241 |
|  delete.SimpleDelete | 19401 | 19401 | 323 | 581 | 1040 |
|  delete.NestedDeleteInUpdate | 23202 | 23202 | 182 | 253 | 2040 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 161 | 60 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 126 | 160 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 59 | 61 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 87 | 122 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 25 | 34 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 55 | 63 | 2174 |
|  query.QueryWhere | 9708 | 9718 | 52 | 52 | 2166 |
|  query.SimpleQueryWithNestedWhere | 9886 | 9896 | 78 | 72 | 2166 |
|  query.Nested | 10096041 | 10096041 | 12080 | 13196 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 11868 | 12991 | 2004000 |
|  query.OrFilterOnRelationships | 43570 | 41053 | 232 | 295 | 1956 |
|  query.OrFilterOnRelationshipsAndNested | 37303 | 35187 | 273 | 369 | 1956 |
| 🟥 query.QueryWithNestedIn | 14744 | 13190 | 68 | 82 | 1664 |
|  query.NestedConnectionWhere | 9834 | 9834 | 67 | 80 | 2174 |
|  query.DeeplyNestedConnectionWhere | 10049 | 9891 | 118 | 112 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19308 | 19098 | 199 | 188 | 1652 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 241 | 502 | 1134 |
|  query.Fulltext | 80 | 80 | 42 | 53 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 66 | 67 | 84 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 85 | 90 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 1459 | 1404 | 222 | 376 | 1040 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 63 | 62 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 108 | 117 | 4516 |
|  sorting.SortWithTopLevelCypher | 3121 | 3121 | 152 | 281 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 94 | 124 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 81 | 118 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 454 | 372 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 328 | 364 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 126 | 175 | 2002 |

</details>
        
Old Schema Generation: 54.145s
Schema Generation: 55.861s
Old Subgraph Schema Generation: 1:06.213 (m:ss.mmm)
Subgraph Schema Generation: 1:07.851 (m:ss.mmm)

File: packages/graphql-toolbox/src/utils/utils.ts [hunk 4]
```diff
@@ -17,36 +17,6 @@
  * limitations under the License.
  */
 
-import CodeMirror from "codemirror";
```
Comment (tbwiss) [inline packages/graphql-toolbox/src/utils/utils.ts:20]: Nice!

File: packages/graphql/src/graphql/directives/relationship.ts [hunk 5]
```diff
@@ -17,7 +17,14 @@
  * limitations under the License.
  */
 
-import { DirectiveLocation, GraphQLBoolean, GraphQLDirective, GraphQLList, GraphQLNonNull, GraphQLString } from "graphql";
+import {
```
Comment (tbwiss) [inline packages/graphql/src/graphql/directives/relationship.ts:20]: I suppose this happened when saving this file at some point. But we shouldn't include it as part of this PR

**Review by tbwiss — COMMENTED** [review]
This is great, bye bye CM v5!

File: packages/graphql/src/graphql/directives/relationship.ts (hunk 5 above)
Comment (mjfwebb) [reply packages/graphql/src/graphql/directives/relationship.ts:20]: whoops!

File: packages/graphql-toolbox/package.json (hunk 1 above)
Comment (mjfwebb) [reply packages/graphql-toolbox/package.json]: :+1: https://github.com/neo4j/graphql/pull/3468/commits/0358d8842c07002c30e2bb60ef771a083b5e5924

File: packages/graphql-toolbox/package.json (hunk 2 above)
Comment (mjfwebb) [reply packages/graphql-toolbox/package.json]: :+1: https://github.com/neo4j/graphql/pull/3468/commits/0358d8842c07002c30e2bb60ef771a083b5e5924

File: packages/graphql-toolbox/src/modules/EditorView/Editor.tsx (hunk 3 above)
Comment (mjfwebb) [reply packages/graphql-toolbox/src/modules/EditorView/Editor.tsx]: :+1: https://github.com/neo4j/graphql/pull/3468/commits/04031407b06957d9abc014d383c6fa1cf830308d

File: packages/graphql/src/graphql/directives/relationship.ts (hunk 5 above)
Comment (mjfwebb) [reply packages/graphql/src/graphql/directives/relationship.ts]: https://github.com/neo4j/graphql/pull/3468/commits/06ee128ff8fd241ca1c9a2d55f2e8b0283a02c08

File: packages/graphql-toolbox/src/modules/EditorView/Editor.tsx [hunk 6]
```diff
@@ -49,17 +49,18 @@ export interface Props {
 }
 
 export const Editor = ({ schema }: Props) => {
+    const theme = useContext(ThemeContext);
```
Comment (tbwiss) [inline packages/graphql-toolbox/src/modules/EditorView/Editor.tsx:52]:
```suggestion
```

File: packages/graphql-toolbox/src/modules/EditorView/GraphQLQueryEditor.tsx [hunk 7]
```diff
@@ -17,147 +17,172 @@
...
-    const ref = useRef<HTMLTextAreaElement | null>(null);
+    const store = useStore();
+    const [value, setValue] = useState<string>();
+
+    // Taken from https://github.com/uiwjs/react-codemirror/blob/master/core/src/useCodeMirror.ts
+    const updateListener = EditorView.updateListener.of((vu: ViewUpdate) => {
+        if (
+            vu.docChanged &&
+            // Fix echoing of the remote changes:
+            // If transaction is market as remote we don't have to call `onChange` handler again
```
Comment (tbwiss) [inline packages/graphql-toolbox/src/modules/EditorView/GraphQLQueryEditor.tsx:78]: Not sure though..
```suggestion
            // If transaction is marked as remote we don't have to call `onChange` handler again
```

File: packages/graphql-toolbox/src/modules/EditorView/Editor.tsx [hunk 8]
```diff
@@ -49,17 +49,18 @@ export interface Props {
...
     const store = useStore();
     const settings = useContext(SettingsContext);
-    const theme = useContext(ThemeContext);
     const [loading, setLoading] = useState<boolean>(false);
     const [showDocs, setShowDocs] = useState<boolean>(false);
-    const refForQueryEditorMirror = useRef<EditorFromTextArea | null>(null);
+    const elementRef = useRef<HTMLDivElement | null>(null);
```
Comment (tbwiss) [inline packages/graphql-toolbox/src/modules/EditorView/Editor.tsx:57]: Can this also be moved to the `GraphQLQueryEditor` file? (In the JSONEditor file the ref is in the same file as the editor)

File: packages/graphql-toolbox/src/modules/EditorView/Editor.tsx (hunk 6 above)
Comment (mjfwebb) [reply packages/graphql-toolbox/src/modules/EditorView/Editor.tsx:52]: Are you sure? It's being used.

Comment (tbwiss) [reply packages/graphql-toolbox/src/modules/EditorView/Editor.tsx:52]: Uups my bad! Ignore this

**Review by tbwiss — APPROVED** [review]
Great work!

**Pull request merged**